src/server/tiering/op_manager.cc

```diff
@@ -0,0 +1,137 @@
+// Copyright 2024, DragonflyDB authors.  All rights reserved.
+// See LICENSE for licensing terms.
+//
+
+#include "server/tiering/op_manager.h"
+
+#include <optional>
+
+#include "base/logging.h"
+
+namespace dfly::tiering {
+
+void OpQueue::Start() {
+  DCHECK(storage_);
+  worker_ = util::MakeFiber(&OpQueue::Loop, this);
+}
+
+void OpQueue::Stop() {
+  cancelled_ = true;
+  waker_.notify();
+  worker_.JoinIfNeeded();
+}
+
+void OpQueue::TEST_Drain() {
+  while (!key_queue_.empty())
+    util::ThisFiber::Yield();
+  Stop();
+}
+
+void OpQueue::Enqueue(std::string_view key, Op op) {
+  auto [it, inserted] = key_ops_.try_emplace(key);
+
+  // DEL can only be followed by SET because we expect the keystore to track item validity
+  DCHECK(it->second.empty() || it->second.back().type != Op::DEL || op.type == Op::SET);
+
+  it->second.push_back(std::move(op));
+
+  if (inserted) {
+    key_queue_.emplace_back(key);
+    waker_.notify();
+  }
+}
+
+void OpQueue::Loop() {
+  while (true) {
+    waker_.await([this] { return !key_queue_.empty() || cancelled_; });
+    if (cancelled_)
+      return;
+
+    const auto& key = key_queue_.front();
+    std::optional<std::string> value;
+
+    // Any write invalidates the stored value, so read only if necessary.
+    // Note: this is a suspension point, key_ops_ can grow!
+    if (key_ops_[key].front().type == Op::GET)
+      value = storage_->Read(key);
+
+    // Atomically fulfill all pending operations
+    bool updated = false;
+    {
+      util::FiberAtomicGuard guard;
+      for (auto& op : key_ops_[key]) {
+        switch (op.type) {
+          case Op::GET:
+            DCHECK(value);  // No GET after DEL
+            break;
+          case Op::DEL:
+            value.emplace();
```

Comment:
`value.emplace()` will fill the optional with an empty string. If there is a `GET` operation following the `DEL` the `DCHECK` won't trigger because the value will contain an empty string. Am I missing something?

Reply:
True, good catch, it should be reset